python/ray/data/grouped_dataset.py

```diff
@@ -249,39 +249,52 @@ def map_groups(
         else:
             sorted_ds = self._dataset.repartition(1)
 
-        def get_key(row):
-            if isinstance(self._key, Callable):
-                return self._key(row)
-            elif isinstance(self._key, str):
-                return row[self._key]
-            else:
+        import numpy as np
+
+        # Returns the keys of the batch in numpy array.
+        # Returns None if the self._key is None.
+        def get_keys(batch, accessor) -> Optional[np.ndarray]:
+            import pandas as pd
+            import pyarrow as pa
+
+            if self._key is None:
                 return None
+            if isinstance(batch, pd.DataFrame) or isinstance(batch, pa.Table):
+                assert isinstance(self._key, str)
+                return batch[self._key].to_numpy()
+            elif isinstance(batch, List):
+                assert callable(self._key)
+                return np.array([self._key(item) for item in accessor.iter_rows()])
```

Comment:
Cool, it's better than `get_keys() -> Optional[np.ndarray]` in `BlockAccessor`, thanks for simplification.